PRIVATE BREAKING: Look up template refs for node info and events on the fly - RFC

evs-chris opened the pull request:
Template refs for indices and keys are currently forwarded along through the fragment hierarchy as it gets created. This seems like basically trading performance for memory, which is fine, but it also introduces a maintenance burden for splicing. That maintenance is not currently being done per #2399 and #2400.

This PR drops the forwarding of refs through and instead looks them up when needed e.g. when an event is fired or when `Ractive.getNodeInfo` is called. I looked briefly at trying to keep the `indexRefs` and `keyRefs` objects up to date when a shuffle happens, but it looked like more code than it's worth for the two places they're actually used. The only potential downside is that gathering the refs for a very very deep template on event firing might add a very slight delay to the event handler being called.

I also added the `keys` to the `event` object because it was free. That brings around the questions of whether there should actually be separate objects for indices and keys, and whether aliases should also be thrown in, since that would be basically free with this approach.

Thoughts?

martypdx commented:
> also introduces a maintenance burden for splicing. That maintenance is not currently being done per #2399 and #2400.

Definitely correct splicing is top priority.

> Thoughts?

The `eventsOrInfo.index.i` api of the event/nodeInfo was always a bit hackish as it required intimate knowledge of the template (though I think the feature needs to be there for nested templates) and requires an index to be defined to get the index (no `@index` equivelent).

What if we move forward with what you have and add a deprecation warning that `.index` and `.key`, and also new api to be `.getIndex( [alias] )` and `.getIndex( [alias] )` respectively.

Then you can write:

```js
this.event.getIndex()  //default to @index !
this.event.getIndex( 'i' ) //gets the index alias defined by i
```

__AND__ it would have the added benefit of being lazy about running `gatherRefs` (and could do so in a more targeted fashion fetching just the index or key being requested).

evs-chris commented:
Interesting! My only concern with the lazy evaluation would be that stashing it for later might cause issues if stuff disappears in the interim. Beyond that, it basically looks like pre-anchored reference resolving from an element, so you could drop the index or key specificity and just make it `get` to find any ref available from that point in the template. See what I did there? :stuck_out_tongue_winking_eye:

Also, I'm not sure how to do a deprecation for property access unless we leave IE8 out shivering in the cold... where it belongs.

martypdx commented:
> My only concern with the lazy evaluation would be that stashing it for later might cause issues if stuff disappears in the interim

Does it need to be cached at all?

> so you could drop the index or key specificity and just make it `get` to find any ref available from that point in the template. See what I did there?

:astonished: 

> Also, I'm not sure how to do a deprecation for property access unless we leave IE8 out shivering in the cold... where it belongs.

Silly :rabbit:, deprecations are for developers not users. We can use `defineProperty` like we do [here](https://github.com/ractivejs/ractive/blob/dev/src/Ractive/construct.js#L37). Which only needs to work on dev machine.

evs-chris commented:
> Does it need to be cached at all?

Not from our end, but I was thinking the dev may stash the event object for some reason. Probably not a terribly reasonable thing to do or worry about.

> :astonished:

Heh. Maybe I'm going off the rails further here, but I think it'd be pretty slick to make the event and node info objects basically a pointer into the data with its own `get`, `set`, `splice`, etc.

> Silly :rabbit:, deprecations are for developers not users.

Well, dang. :smile:

sabob commented:
Thanks Chris! I was looking for where indexrefs were updated, now I know, they aren't ;-)

To fix the keypaths for #2400 should updatePostShuffle update the "fragment.key" as well as "fragment.context.key" to the index value? Or it's not so straightforward?

evs-chris commented:
@sabob, no, the `getNodeInfo` call should just be pulling the keypath on the fly too. I misread #2400, so I thought I had it covered. On the plus side, I think we can eliminate a little more memory footprint :smile:

evs-chris commented:
Since we already have everything we need for node info in the fragment, I just dropped the context and keypath stuff out of `_ractive`. It seems like more unnecessary work during an already busy process to keep those up to date during rebind. I also dropped the ractive reference because it is readily available from the fragment. Is that fair?

If this is all good, I think I'll punt deprecating and replacing the index and key maps to another PR.

martypdx commented:
> Heh. Maybe I'm going off the rails further here, but I think it'd be pretty slick to make the event and node info objects basically a pointer into the data with its own get, set, splice, etc.

Actually that was the "astonished" emoji - meaning it's a great idea

evs-chris commented:
:smile: It also neatly solves 2/3s of the issue I was trying to address with anchored set, splice, etc. I'm imagining `<button on-click="event.splice('../', @index, 1)">`, but I don't know that that's entirely reasonable.

martypdx commented:
Just hang the methods of `@keypath` (or add `@context` if you prefer):

`<button on-click="@keypath.splice('../', @index, 1)">`

sabob commented:
Should we also return the "binding" property as part of nodeInfo?

I've seen binding used in decorators such as:
https://github.com/Prezent/ractive-decorators-select2/blob/master/ractive-decorators-select2.js

where the binding and keypath is looked up via:
``` var binding = node._ractive.binding;
var keypath = binding.keypath ? binding.keypath : binding.model.key;```
